Doc: multithreading: explicit why GDALDataset is not thread-safe (#10097)
OSGeo/gdal: doc/source/user/multithreading.rst

:cpp:class:`GDALDataset` instance (each thread should instead manipulate a
distinct GDALDataset). Similarly for a GDALDataset owning several :cpp:class:`OGRLayer`.

The reason is that most implementations of GDALDataset or GDALRasterBand

> are not state-less. A GDALDataset typically owns a file handle,

Reviewer: How about the following, removing the double negative?

```suggestion
are stateful. A GDALDataset typically owns a file handle,
```

and performs seek/read operations on it, thus not allowing concurrent access.
Block cache related structures for a given GDALDataset are not thread-safe.
Drivers also often implement lazy initialization strategies to access various
metadata which are resolved only the first time the method to access them is
invoked. Drivers may also rely on third-party libraries that expose objects

> that are not tread-safe.

Reviewer 2: ```suggestion
that are not thread-safe.
```

Those restrictions apply to the C and C++ ABI, and all languages bindings (unless
they would take special precautions to serialize calls)

GDAL block cache and multi-threading
------------------------------------

> The current design of the GDAL raster block cache makes it appropriate to
> read several datasets from several threads. However performance issues may

Reviewer: ```suggestion
The current design of the GDAL raster block cache allows concurrent reads of several datasets. However performance issues may
```
